### ISSUE TITLE
`coder_app` order  is not being honored in the UI, but `slug` is

### ISSUE THREAD

**Issue by michaelbrewer** [coder/coder#12467]
According to the coder docs : [ UI Resource Ordering- Applications](https://coder.com/docs/v2/latest/templates/resource-ordering#applications) the follow code should make the order:

1. gyft
2. fiserv
3. git

```terraform
resource "coder_app" "six" {
  agent_id     = coder_agent.main.id
  slug         = "a1"
  display_name = "Git project (order = 3, slug = a1)"
  order        = 3
  icon         = "/icon/git.svg"
  url          = "${var.dotfiles_uri}"
  external     = true
}

resource "coder_app" "five" {
  agent_id     = coder_agent.main.id
  slug         = "a2"
  display_name = "Fiserv Site (order = 2, slug = a2)"
  order        = 2
  icon         = "/icon/apple-black.svg"
  url          = "http://fiserv.com"
  external     = true
}

resource "coder_app" "seven" {
  agent_id     = coder_agent.main.id
  slug         = "a3"
  display_name = "Gyft site (order = 1, slug = a3)"
  order        = 1
  icon         = "/icon/apache-guacamole.svg"
  url          = "http://www.gyft.com"
  external     = true
}
```

But the displayed order is based on the slug name

<img width="1254" alt="Screenshot 2024-03-07 at 6 02 42 PM" src="https://github.com/coder/coder/assets/5442469/4f0d1834-fb2f-4451-8c4b-7f601ed33fe5">

---

Db query looks at `slug` for the order?

https://github.com/coder/coder/blob/d2a74cf547ddb3f66b3a4f97007037fb7ecce166/coderd/database/queries/workspaceapps.sql#L5

I do see that the sorting is happening here, but i wonder if that is what the UI api call is using?

https://github.com/coder/coder/blob/d2a74cf547ddb3f66b3a4f97007037fb7ecce166/coderd/database/db2sdk/db2sdk.go#L454-L461

And the sorting only gets called on the POST api call 

https://github.com/coder/coder/blob/d2a74cf547ddb3f66b3a4f97007037fb7ecce166/coderd/workspaces.go#L345

and not the GET api call 

https://github.com/coder/coder/blob/d2a74cf547ddb3f66b3a4f97007037fb7ecce166/coderd/workspaces.go#L245-L248

**Comment by mtojek** [comment]
1. What is the template provider version you're using?
2. What is the release version of coderd?

**Comment by michaelbrewer** [comment]
The latest versions of each

1. coder/coder v0.18.0 (signed by a HashiCorp partner, key ID 93C75807601AA0EC)
2. Coder v2.8.4+2d37eb4

**Comment by michaelbrewer** [comment]
@mtojek  - in this test i am just using the latest 

```terraform
terraform {
  required_providers {
    coder = {
      source = "coder/coder"
    }
  }
}
// ...

resource "coder_agent" "main" {
  arch                   = data.coder_provisioner.me.arch
  os                     = "linux"
  startup_script         = <<-EOT
    set -e
  EOT

  display_apps {
    vscode          = false
    vscode_insiders = false
    web_terminal    = false
    ssh_helper      = false
  }
}

resource "coder_app" "six" {
  agent_id     = coder_agent.main.id
  slug         = "a1"
  display_name = "Git project (order = 3, slug = a1)"
  order        = 3
  icon         = "/icon/git.svg"
  url          = "https://github.com/"
  external     = true
}

resource "coder_app" "five" {
  agent_id     = coder_agent.main.id
  slug         = "a2"
  display_name = "Fiserv Site (order = 2, slug = a2)"
  order        = 2
  icon         = "/icon/apple-black.svg"
  url          = "https://fiserv.com"
  external     = true
}

resource "coder_app" "seven" {
  agent_id     = coder_agent.main.id
  slug         = "a3"
  display_name = "Gyft site (order = 1, slug = a3)"
  order        = 1
  icon         = "/icon/apache-guacamole.svg"
  url          = "https://www.gyft.com"
  external     = true
}

// ... rest

```

**Comment by mtojek** [comment]
I'm afraid I can't reproduce it. It works as intended on my side. I will check against the latest version.

> I do see that the sorting is happening here, but i wonder if that is what the UI api call is using?

It should be called: `workspaceByOwnerAndName` -> `workspaceData` -> `convertWorkspaceBuilds` -> `convertWorkspaceBuild` -> `db2sdk.Apps`.

You can check a few other things:
1. Ensure you have `order` property defined for every `coder_app`.
2. Check `workspace_apps` table if `display_order` column is populated.

**Comment by michaelbrewer** [comment]
> It should be called: workspaceByOwnerAndName -> workspaceData -> convertWorkspaceBuilds -> convertWorkspaceBuild -> db2sdk.Apps. 

Oh, ok i see that. Strange.

1. Ensure you have order property defined for every coder_app. ✅ 
2. Make sure to update the workspace once the new template revision (with order) is pushed. ✅ 
3. Check `workspace_apps` table if `display_order` column is populated. 🚫 

```text
coder=> select * from workspace_apps where agent_id = '985cfb72-8c1e-4afe-ac23-e86a68f5bd5f' limit 100;
-[ RECORD 1 ]---------+-------------------------------------
id                    | f368f9f7-7048-4b38-bbd4-d03fcb26d5cf
created_at            | 2024-03-08 09:57:00.32657+00
agent_id              | 985cfb72-8c1e-4afe-ac23-e86a68f5bd5f
display_name          | Git project (order = 3, slug = a1)
icon                  | /icon/git.svg
command               |
url                   | https://github.com/
healthcheck_url       |
healthcheck_interval  | 0
healthcheck_threshold | 0
health                | disabled
subdomain             | f
sharing_level         | owner
slug                  | a1
external              | t
-[ RECORD 2 ]---------+-------------------------------------
id                    | 5c119c3e-021d-466e-b1c8-66cce838277b
created_at            | 2024-03-08 09:57:00.326961+00
agent_id              | 985cfb72-8c1e-4afe-ac23-e86a68f5bd5f
display_name          | Fiserv Site (order = 2, slug = a2)
icon                  | /icon/apple-black.svg
command               |
url                   | https://fiserv.com
healthcheck_url       |
healthcheck_interval  | 0
healthcheck_threshold | 0
health                | disabled
subdomain             | f
sharing_level         | owner
slug                  | a2
external              | t
-[ RECORD 3 ]---------+-------------------------------------
id                    | 50c81fd5-9d5e-4f31-ba5f-0557634cdbdb
created_at            | 2024-03-08 09:57:00.327296+00
agent_id              | 985cfb72-8c1e-4afe-ac23-e86a68f5bd5f
display_name          | Gyft site (order = 1, slug = a3)
icon                  | /icon/apache-guacamole.svg
command               |
url                   | https://www.gyft.com
healthcheck_url       |
healthcheck_interval  | 0
healthcheck_threshold | 0
health                | disabled
subdomain             | f
sharing_level         | owner
slug                  | a3
external              | t
```

**Comment by michaelbrewer** [comment]
In fact there is no `display_order` column?

```
coder=> \d workspace_apps
                                       Table "public.workspace_apps"
        Column         |           Type           | Collation | Nullable |             Default
-----------------------+--------------------------+-----------+----------+----------------------------------
 id                    | uuid                     |           | not null |
 created_at            | timestamp with time zone |           | not null |
 agent_id              | uuid                     |           | not null |
 display_name          | character varying(64)    |           | not null |
 icon                  | character varying(256)   |           | not null |
 command               | character varying(65534) |           |          |
 url                   | character varying(65534) |           |          |
 healthcheck_url       | text                     |           | not null | ''::text
 healthcheck_interval  | integer                  |           | not null | 0
 healthcheck_threshold | integer                  |           | not null | 0
 health                | workspace_app_health     |           | not null | 'disabled'::workspace_app_health
 subdomain             | boolean                  |           | not null | false
 sharing_level         | app_sharing_level        |           | not null | 'owner'::app_sharing_level
 slug                  | text                     |           | not null |
 external              | boolean                  |           | not null | false
Indexes:
    "workspace_apps_pkey" PRIMARY KEY, btree (id)
    "workspace_apps_agent_id_slug_idx" UNIQUE CONSTRAINT, btree (agent_id, slug)
Foreign-key constraints:
    "workspace_apps_agent_id_fkey" FOREIGN KEY (agent_id) REFERENCES workspace_agents(id) ON DELETE CASCADE
```

**Comment by mtojek** [comment]
Yes, this is unusual :)

Could you please double-check the build-info: `/api/v2/buildinfo`?

**Comment by michaelbrewer** [comment]
Buildinfo is `,"version":"v2.8.4+2d37eb4"` but my schema was not upgraded 

```bash
coder=> select * from schema_migrations;
-[ RECORD 1 ]
version | 187
dirty   | f
```

**Comment by michaelbrewer** [comment]
I did an upgrade via `curl -L https://coder.com/install.sh | sh` but i am not using the embedded db, so i am not sure how it finds the db to upgrade.

**Comment by mtojek** [comment]
In theory, Coder should apply the next migrations on startup. It would be great to check for any errors in application logs.

Could you please restart `coder` service and see if something bumps up in logs or if migrations are executed?

**Comment by michaelbrewer** [comment]
@mtojek right now it is not running as a service from the commandline (i don't see how to set HTTP_PROXY via a service)

**Comment by michaelbrewer** [comment]
@mtojek - and i see no errors in startup

**Comment by mtojek** [comment]
Alright, I think I know what is happening.

TL;DR You need to wait for the next release `v2.9.0` or switch to the unstable latest.

The release `v2.8.4` was a non-usual one as it contains a cherry-picked security patch. It confirms why you see only [187 migrations](https://github.com/coder/coder/tree/v2.8.4/coderd/database/migrations), and I can see more. The best option now is to wait until we cut a new release or switch to the latest commit, which might not be super stable.

**Comment by michaelbrewer** [comment]
ok thanks. For looking into it and make sense why it was not working :)

**Comment by michaelbrewer** [comment]
(So i guess the risk here is that the docs is actually ahead of what is available - Maybe the docs can be versioned to?)

**Comment by michaelbrewer** [comment]
OK, cool i will close this. Maybe it would have for the coder server can point to the related docs for the deployed version?

**Comment by michaelbrewer** [comment]
Currently "Documentation" link points to `https://coder.com/docs/v2/latest`, maybe a feature request i can put in. 

(It did find that SUPPORT_LINKS flag actually removed the coder links :()
